exonum/src/events/codec.rs

```diff
@@ -80,13 +79,30 @@ impl Decoder for MessagesCodec {
             );
         }
 
-        // Read message
-        if buf.len() >= total_len {
-            let data = buf.split_to(total_len).to_vec();
-            let raw = RawMessage::new(MessageBuffer::from_vec(data));
-            return Ok(Some(raw));
+        if total_len > buf.len() {
+            bail!(
+                "Received malicious message with wrong \
+                 total_len: {}, expected message length {}",
+                total_len,
+                buf.len()
+            );
+        }
+
+        let data = buf.split_to(total_len).to_vec();
+        let raw = RawMessage::new(MessageBuffer::from_vec(data));
+        Ok(Some(raw))
+    }
+
+    fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        if let Some(frame) = try!(self.decode(buf)) {
```

Comment:
Nit: Judging by [docs](https://doc.rust-lang.org/std/macro.try.html), using `?` operator instead of `try!` is preferred.

Also, I personally would find the following implementation more obvious:

```rust
let maybe_message = self.decode(buf)?;
if maybe_message.is_none() && !buf.is_empty() {
    trace!(...);
}
maybe_message
```